File: server-go/device_rate_test.go
package main

import "testing"

// The per-device verification rate gate. A precondition on token issuance, not
// weighted evidence: see deviceVerificationsPerMinute in scoring.go.

func deviceSignals(instance string) map[string]interface{} {
	sig := map[string]interface{}{
		"behavioral": map[string]interface{}{
			"totalPoints": 60.0, "trajectoryLength": 400.0, "approachPoints": 12.0,
			"approachDirectness": 0.4, "microTremorScore": 0.5, "velocityVariance": 0.5,
		},
		"environmental": map[string]interface{}{"automationFlags": map[string]interface{}{}},
	}
	if instance != "" {
		sig["meta"] = map[string]interface{}{"sessionId": instance}
	}
	return sig
}

func TestDeviceRateGateWithholdsTheEleventhVerification(t *testing.T) {
	e := NewScoringEngine("test-secret")
	const ip = "203.0.113.77"
	for i := 0; i < deviceVerificationsPerMinute; i++ {
		r := e.VerifyWithHeaders(deviceSignals("page-a"), ip, "site", "ua", nil, "", "", false, nil, TokenBinding{})
		if r.Reason == "rate_limited" {
			t.Fatalf("verification %d of %d was rate limited early", i+1, deviceVerificationsPerMinute)
		}
	}
	r := e.VerifyWithHeaders(deviceSignals("page-a"), ip, "site", "ua", nil, "", "", false, nil, TokenBinding{})
	if r.Success || r.Reason != "rate_limited" {
		t.Errorf("verification %d should be withheld as rate_limited, got success=%v reason=%q",
			deviceVerificationsPerMinute+1, r.Success, r.Reason)
	}
	if !hasReasonContaining(r.Detections, "for this device") {
		t.Errorf("expected the per-device rate detection to be recorded, got %+v", r.Detections)
	}

	// Another page instance on the same address and device has its own budget:
	// identical machines behind one NAT must not share one.
	other := e.VerifyWithHeaders(deviceSignals("page-b"), ip, "site", "ua", nil, "", "", false, nil, TokenBinding{})
	if other.Reason == "rate_limited" {
		t.Errorf("a different widget instance must not inherit the exhausted budget")
	}
}

func TestDeviceRateGateNeedsAWidgetInstance(t *testing.T) {
	// A client that reports no instance id — an older widget — is not gated,
	// because the alternative key (address + fingerprint) is shared by every
	// identical machine behind one address.
	e := NewScoringEngine("test-secret")
	for i := 0; i <= deviceVerificationsPerMinute+2; i++ {
		r := e.VerifyWithHeaders(deviceSignals(""), "203.0.113.78", "site", "ua", nil, "", "", false, nil, TokenBinding{})
		if r.Reason == "rate_limited" {
			t.Fatalf("no instance id must mean no device gate, got rate_limited on call %d", i+1)
		}
	}
}

func TestWidgetInstanceIsBounded(t *testing.T) {
	long := make([]byte, 500)
	for i := range long {
		long[i] = 'x'
	}
	sig := map[string]interface{}{"meta": map[string]interface{}{"widgetId": string(long)}}
	if got := widgetInstance(sig); len(got) != 64 {
		t.Errorf("client-supplied id must be bounded before it becomes a key, got length %d", len(got))
	}
	if widgetInstance(map[string]interface{}{}) != "" {
		t.Error("no meta means no instance")
	}
}

File: server-node/device-rate.test.js
'use strict';

// The per-device verification rate gate: a precondition on token issuance, not
// weighted evidence. Mirrors server-go/device_rate_test.go.

const { test } = require('node:test');
const assert = require('node:assert/strict');
const { createScoringEngine } = require('./index');
const { widgetInstance, DEVICE_VERIFICATIONS_PER_MINUTE } = require('./engine');

const signals = (instance) => ({
  behavioral: { totalPoints: 60, trajectoryLength: 400, approachPoints: 12,
    approachDirectness: 0.4, microTremorScore: 0.5, velocityVariance: 0.5 },
  environmental: { automationFlags: {} },
  ...(instance ? { meta: { sessionId: instance } } : {}),
});
const headers = { accept: 'text/html', 'accept-language': 'en-US' };

test('the eleventh verification from one page instance is withheld as rate_limited', () => {
  const engine = createScoringEngine({ secret: 'test-secret' });
  for (let i = 0; i < DEVICE_VERIFICATIONS_PER_MINUTE; i++) {
    const r = engine.verify(signals('page-a'), '203.0.113.77', 'site', 'ua', headers);
    assert.notEqual(r.reason, 'rate_limited', `verification ${i + 1} was rate limited early`);
  }
  const r = engine.verify(signals('page-a'), '203.0.113.77', 'site', 'ua', headers);
  assert.equal(r.success, false);
  assert.equal(r.reason, 'rate_limited');
  assert.ok(r.detections.some((d) => /for this device/.test(d.reason)));

  // Another page instance on the same address and device has its own budget.
  const other = engine.verify(signals('page-b'), '203.0.113.77', 'site', 'ua', headers);
  assert.notEqual(other.reason, 'rate_limited');
});

test('no widget instance means no device gate', () => {
  const engine = createScoringEngine({ secret: 'test-secret' });
  for (let i = 0; i <= DEVICE_VERIFICATIONS_PER_MINUTE + 2; i++) {
    const r = engine.verify(signals(''), '203.0.113.78', 'site', 'ua', headers);
    assert.notEqual(r.reason, 'rate_limited');
  }
});

test('the instance id is bounded before it becomes a key', () => {
  assert.equal(widgetInstance({ meta: { widgetId: 'x'.repeat(500) } }).length, 64);
  assert.equal(widgetInstance({}), '');
  assert.equal(widgetInstance({ meta: { sessionId: 42 } }), '');
});
